src/QueryBuilder/ScoringQueryBuilder.php

```diff
@@ -0,0 +1,174 @@
+<?php
+
+namespace Pucene\Bundle\DoctrineBundle\QueryBuilder;
+
+use Doctrine\ORM\EntityManagerInterface;
+use Doctrine\ORM\Query\Expr\Join;
+use Doctrine\ORM\QueryBuilder;
+use Pucene\Bundle\DoctrineBundle\Entity\Document;
+
+class ScoringQueryBuilder
+{
+    /**
+     * @var EntityManagerInterface
+     */
+    private $entityManager;
+
+    /**
+     * @var QueryBuilder
+     */
+    private $queryBuilder;
+
+    /**
+     * @var string[]
+     */
+    private $parts = [];
+
+    /**
+     * @var string[]
+     */
+    private $joins = [];
+
+    /**
+     * @var int
+     */
+    private $docCount;
+
+    /**
+     * @var int[]
+     */
+    private $docCountPerTerm = [];
+
+    /**
+     * @param EntityManagerInterface $entityManager
+     * @param QueryBuilder $queryBuilder
+     */
+    public function __construct(EntityManagerInterface $entityManager, QueryBuilder $queryBuilder)
+    {
+        $this->entityManager = $entityManager;
+        $this->queryBuilder = $queryBuilder;
+
+        $this->queryBuilder->from(Document::class, 'innerDocument')
+            ->where('innerDocument.id = document.id')
+            ->setMaxResults(1);
+    }
+
+    public function open()
+    {
+        $this->parts[] = '(';
+    }
+
+    public function multiply()
+    {
+        $this->parts[] = '*';
+    }
+
+    public function addition()
+    {
+        $this->parts[] = '+';
+    }
+
+    public function queryNorm(array $tokens)
+    {
+        $sum = 0;
+        foreach ($tokens as $token) {
+            $sum += pow($this->calculateInverseDocumentFrequency($token->getTerm()), 2);
+        }
+
+        $this->parts[] = (1.0 / sqrt($sum));
+    }
+
+    public function termFrequency($term)
+    {
+        $this->parts[] = 'COALESCE(SQRT(' . $this->joinTerm($term) . '.frequency),0)';
```

Comment:
i will use the `Expr\Func` for `COALESCE` directly. because the `COALESCE` is implemented in mysql/postgre internally and the `CASE` will be executed as case (which is slower) and for case there is also no expr